envoy_qatzip/compressor/qatzip_compressor_impl.cc

```diff
@@ -0,0 +1,101 @@
+#include "envoy_qatzip/compressor/qatzip_compressor_impl.h"
+
+#include <memory>
+
+#include "common/common/assert.h"
+
+#include "absl/container/fixed_array.h"
+
+namespace Envoy {
+namespace Compressor {
+
+QatzipCompressorImpl::QatzipCompressorImpl() : QatzipCompressorImpl(4096) {}
+
+QatzipCompressorImpl::QatzipCompressorImpl(size_t chunk_size)
+    : initialized_{false}, chunk_size_{chunk_size}, avail_out_{chunk_size},
+      chunk_char_ptr_(new unsigned char[chunk_size]), session_{}, stream_{} {
+  stream_.out = chunk_char_ptr_.get();
+}
+
+QatzipCompressorImpl::~QatzipCompressorImpl() {
+  if (initialized_) {
+    qzEndStream(&session_, &stream_);
+    qzTeardownSession(&session_);
+    qzClose(&session_);
+  }
+}
+
+void QatzipCompressorImpl::init(unsigned int compression_level, unsigned int hardware_buffer_size,
+                                unsigned int stream_buffer_size,
+                                unsigned int input_size_threshold) {
+  QzSessionParams_T params;
+
+  ASSERT(initialized_ == false);
+
+  int status = qzGetDefaults(&params);
+  RELEASE_ASSERT(status == QZ_OK, "failed to initialize hardware");
+  params.comp_lvl = compression_level;
+  params.hw_buff_sz = hardware_buffer_size;
+  params.strm_buff_sz = stream_buffer_size;
+  params.input_sz_thrshold = input_size_threshold;
+
+  status = qzInit(&session_, params.sw_backup);
+  RELEASE_ASSERT(status == QZ_OK || status == QZ_DUPLICATE, "failed to initialize hardware");
+  status = qzSetupSession(&session_, &params);
```

Comment:
have you tried two filter chains with different settings?

Reply:
No, I haven't. I assumed `QzSession_T` contains state data specific to a session(stream). I check in QATzip's source code it's not global and multiple sessions with different params can coexist at the same time. Then QAT handles chunks of data with the session params included in the chunks, so it's stateless.

But yeah, would be nice to test it, not to assume...

Reply:
+1